File: boj/1633.py
import sys

chart=[]
for line in sys.stdin:
    line = line.strip()
    if not line:  # 빈 줄이면 중단
        break
    a, b = map(int, line.split())
    chart.append((a,b))
    #print(a, b)

#print(chart)

dp=[[[0 for i in range(16)] for j in range(16)] for k in range(len(chart))]

def cal(index, w, b):
    global dp, chart
    if index == -1:
        return 0
    # 흰 선택
    if dp[index][w][b]:
        return dp[index][w][b]
    white=0
    black=0
    none=0
    if w > 0:
        white=cal(index-1, w-1, b)+chart[index][0]
    # 검 선택
    if b > 0:
        black=cal(index-1, w, b-1)+chart[index][1]

    # 안 선택
    none=cal(index-1, w, b)
    dp[index][w][b]=max(white, black, none)
    return dp[index][w][b]


print(cal(len(chart)-1, 15, 15))

File: boj/1633.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class Main {

    static int n;
    static int[][] chart;
    static int[][][] dp;

    private static int dp(int idx, int white_cnt, int black_cnt){
        if(idx == -1 && white_cnt == 0 && black_cnt == 0){
            return 0;
        }
        if(idx == -1) {
            return -1;
        }
        if(dp[idx][white_cnt][black_cnt] != -1){
            return dp[idx][white_cnt][black_cnt];
        }


        int ret = dp(idx-1, white_cnt, black_cnt);
        if(white_cnt > 0){
            ret = Math.max(ret, dp(idx-1, white_cnt - 1, black_cnt) + chart[idx][0]);
        }

        if(black_cnt > 0){
            ret = Math.max(ret, dp(idx-1, white_cnt, black_cnt - 1) + chart[idx][1]);
        }
        dp[idx][white_cnt][black_cnt] = ret;

        return ret;
    }

    private static void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        chart = new int[1000][2];
        String input = "";
        while((input = br.readLine()) != null) {
            if (input.isEmpty()) break;
            String[] t = input.split(" ");
            int black = Integer.parseInt(t[0]);
            int white = Integer.parseInt(t[1]);
            chart[n][0] = black;
            chart[n][1] = white;
            n+=1;
        }
        dp=new int [n][16][16];

        for(int i = 0; i < n; i++){
            for(int  j = 0; j < 16; j++){
                Arrays.fill(dp[i][j], -1);
            }
        }

        System.out.println(dp(n-1, 15, 15));


    }


    public static void main(String[] args) throws Exception {
        solution();
    }
}

